core/src/main/java/com/sequenceiq/cloudbreak/core/flow/ClusterBootstrapperErrorHandler.java

```diff
@@ -0,0 +1,112 @@
+import com.sequenceiq.cloudbreak.domain.Resource;
+import com.sequenceiq.cloudbreak.domain.Stack;
+import com.sequenceiq.cloudbreak.orchestrator.CloudbreakOrchestratorException;
+import com.sequenceiq.cloudbreak.orchestrator.ContainerOrchestrator;
+import com.sequenceiq.cloudbreak.orchestrator.Node;
+import com.sequenceiq.cloudbreak.repository.HostMetadataRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceGroupRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceMetaDataRepository;
+import com.sequenceiq.cloudbreak.repository.ResourceRepository;
+import com.sequenceiq.cloudbreak.service.stack.resource.DeleteContextObject;
+import com.sequenceiq.cloudbreak.service.stack.resource.ResourceBuilder;
+import com.sequenceiq.cloudbreak.service.stack.resource.ResourceBuilderInit;
+
+@Component
+public class ClusterBootstrapperErrorHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClusterBootstrapperErrorHandler.class);
+
+    @Autowired
+    private ResourceRepository resourceRepository;
+
+    @Autowired
+    private InstanceMetaDataRepository instanceMetaDataRepository;
+
+    @Autowired
+    private InstanceGroupRepository instanceGroupRepository;
+
+    @Autowired
+    private HostMetadataRepository hostMetadataRepository;
+
+    @javax.annotation.Resource
+    private Map<CloudPlatform, List<ResourceBuilder>> instanceResourceBuilders;
+
+    @javax.annotation.Resource
+    private Map<CloudPlatform, ResourceBuilderInit> resourceBuilderInits;
+
+    public void handleError(ContainerOrchestrator orchestrator, Stack stack, Set<Node> nodes) throws CloudbreakOrchestratorException {
+        InstanceGroup gateway = stack.getGatewayInstanceGroup();
+        InstanceMetaData gatewayInstance = gateway.getInstanceMetaData().iterator().next();
+        List<String> allAvailableNode = orchestrator.getAllAvailableNode(gatewayInstance.getPublicIp(), nodes);
+        List<Node> missingNodes = selectMissingNodes(nodes, allAvailableNode);
+        if (missingNodes.size() > 0) {
+            LOGGER.info(String.format("Bootstrap failed on some nodes rollbak logic started on %s node.", missingNodes.size()));
+            for (Node missingNode : missingNodes) {
+                InstanceMetaData hostInStack =
+                        instanceMetaDataRepository.findByPrivateAndPublicAddress(stack.getId(), missingNode.getPrivateIp(), missingNode.getPublicIp());
+                instanceMetaDataRepository.delete(hostInStack.getId());
+                LOGGER.info(String.format("InstanceMetadata deleted with %s id and %s name.", hostInStack.getId(), hostInStack.getInstanceId()));
+                Set<Resource> resources = resourceRepository.findByStackIdAndName(stack.getId(), hostInStack.getInstanceId());
+                for (Resource resource : resources) {
+                    try {
+                        deleteResourceandDependencies(stack, hostInStack, resource);
```

Comment:
I think it would be better if the cluster creation wouldn't have to wait for the termination of these nodes. If the validation (can we still launch the cluster) is done synchronously than the resource termination can be done async (e.g.: in a new flow phase), so the cluster creation can go on while those resources are terminated.